File: android/app/src/test/java/com/easybc/planner/sync/shared/ProfileLabelsTest.kt
package com.easybc.planner.sync.shared

import com.easybc.planner.sync.shared.ProfileRecord
import com.easybc.planner.sync.shared.SharedSyncState
import com.keyneom.synckit.sharing.SharingRole
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

class ProfileLabelsTest {
  private val profiles = listOf(
      ProfileRecord(
          datasetId = PRIMARY_DATASET_ID,
          ownerEmail = "mom@example.com",
          folderName = "EasyBC — mom@example.com",
          role = SharingRole.OWNER.name.lowercase(),
          trustedOwnerKeyId = "key-1",
      ),
      ProfileRecord(
          datasetId = "daughter",
          ownerEmail = "mom@example.com",
          folderName = "EasyBC — mom@example.com",
          displayName = "Daughter",
          role = SharingRole.OWNER.name.lowercase(),
          trustedOwnerKeyId = "key-1",
      ),
  )

  private val state = SharedSyncState(
      rpId = "keyneom.github.io",
      ownerEmail = "mom@example.com",
      activeProfileKey = "mom@example.com/primary",
      profiles = profiles,
  )

  @Test
  fun slugifyDatasetId_normalizesNames() {
      assertEquals("daughter", slugifyDatasetId("Daughter"))
      assertEquals("sarah-s-cycle", slugifyDatasetId("Sarah's cycle"))
  }

  @Test
  fun uniqueOwnedDatasetId_appendsSuffixOnCollision() {
      assertEquals("daughter-2", uniqueOwnedDatasetId("Daughter", "mom@example.com", profiles))
  }

  @Test
  fun newOwnedDatasetId_isOpaque() {
      assertEquals(
          "p-12345678-1234-1234-1234-123456789abc",
          newOwnedDatasetId(listOf("primary")) {
              "12345678-1234-1234-1234-123456789abc"
          },
      )
  }

  @Test
  fun profileDisplayLabel_usesDisplayNameForOwnedProfiles() {
      assertEquals("My data", profileDisplayLabel(state, profiles[0]))
      assertEquals("Daughter", profileDisplayLabel(state, profiles[1]))
  }

  @Test
  fun profileDisplayLabel_usesLocalOverrideAndDisambiguatesByOwner() {
      val first = profiles[1].copy(ownerEmail = "first@example.com", localDisplayName = "Alex")
      val second = profiles[1].copy(ownerEmail = "second@example.com", localDisplayName = "Alex")
      val value = state.copy(profiles = listOf(first, second))
      assertEquals("Alex", profileDisplayLabel(value, first))
      assertEquals("Alex — first@example.com", disambiguatedProfileLabel(value, first))
      assertEquals("Alex — second@example.com", disambiguatedProfileLabel(value, second))
  }

  @Test
  fun slugifyDatasetId_rejectsPrimary() {
      assertThrows(IllegalArgumentException::class.java) {
          slugifyDatasetId("primary")
      }
  }

  @Test
  fun duplicatePrimaryDatasetIds_areScopedByOwner() {
      val shared = ProfileRecord(
          datasetId = PRIMARY_DATASET_ID,
          ownerEmail = "other@example.com",
          folderName = "EasyBC — other@example.com",
          role = SharingRole.WRITER.name.lowercase(),
          trustedOwnerKeyId = "key-2",
          needsInitialLoad = true,
      )
      val value = state.copy(profiles = profiles + shared)
      assertEquals(profiles[0], findProfile(value, "mom@example.com/primary"))
      assertEquals(shared, findProfile(value, "other@example.com/primary"))
      assertTrue(shouldLoadRemoteBeforePublish(shared))
  }

  @Test
  fun meaningfulLocalData_requiresPreservationBeforeJoin() {
      assertFalse(hasMeaningfulSharedData(emptySharedPayload()))
      assertTrue(
          hasMeaningfulSharedData(
              emptySharedPayload().copy(
                  planner = emptySharedPayload().planner.copy(configured = true),
              ),
          ),
      )
  }

  @Test
  fun localProfile_isNamedAndNeverRequiresRemoteLoad() {
      val local = ProfileRecord(
          datasetId = "profile",
          ownerEmail = "local-123",
          folderName = "",
          displayName = "Offline journal",
          role = SharingRole.OWNER.name.lowercase(),
          trustedOwnerKeyId = "",
          syncMode = "local",
      )
      assertTrue(isLocalProfile(local))
      assertFalse(shouldLoadRemoteBeforePublish(local))
      assertEquals("Offline journal", profileDisplayLabel(state.copy(profiles = listOf(local)), local))
  }
}

File: web/src/sync/profileLabels.test.ts
import { describe, expect, it } from "vitest";
import {
  profileDisplayLabel,
  disambiguatedProfileLabel,
  newOwnedDatasetId,
  slugifyDatasetId,
  uniqueOwnedDatasetId,
} from "./profileLabels";
import { findProfile, type ProfileRecord, type SharedSyncState } from "./sharedTypes";

const state = (profiles: ProfileRecord[]): SharedSyncState => ({
  schemaVersion: 1,
  rpId: "keyneom.github.io",
  ownerEmail: "mom@example.com",
  activeProfileKey: "mom@example.com/primary",
  profiles,
});

const owned = (datasetId: string, displayName?: string): ProfileRecord => ({
  datasetId,
  ownerEmail: "mom@example.com",
  folderName: "EasyBC — mom@example.com",
  role: "owner",
  trustedOwnerKeyId: "key-1",
  ...(displayName ? { displayName } : {}),
});

describe("slugifyDatasetId", () => {
  it("slugifies display names", () => {
    expect(slugifyDatasetId("Daughter")).toBe("daughter");
    expect(slugifyDatasetId("Sarah's cycle")).toBe("sarah-s-cycle");
  });

  it("rejects reserved primary slug", () => {
    expect(() => slugifyDatasetId("primary")).toThrow();
  });
});

describe("uniqueOwnedDatasetId", () => {
  it("returns base slug when unused", () => {
    expect(uniqueOwnedDatasetId("Daughter", "mom@example.com", [owned("primary")])).toBe(
      "daughter",
    );
  });

  it("appends suffix on collision", () => {
    expect(
      uniqueOwnedDatasetId("Daughter", "mom@example.com", [
        owned("primary"),
        owned("daughter"),
      ]),
    ).toBe("daughter-2");
  });
});

describe("newOwnedDatasetId", () => {
  it("creates an opaque id independent of the display label", () => {
    expect(
      newOwnedDatasetId(["primary"], () => "12345678-1234-1234-1234-123456789abc"),
    ).toBe("p-12345678-1234-1234-1234-123456789abc");
  });
});

describe("profileDisplayLabel", () => {
  it("labels primary owned profile as My data", () => {
    expect(profileDisplayLabel(state([owned("primary")]), owned("primary"))).toBe("My data");
  });

  it("uses displayName for owned profiles", () => {
    expect(
      profileDisplayLabel(state([owned("daughter", "Daughter")]), owned("daughter", "Daughter")),
    ).toBe("Daughter");
  });

  it("uses folder name for shared profiles", () => {
    const shared: ProfileRecord = {
      datasetId: "primary",
      ownerEmail: "other@example.com",
      folderName: "EasyBC — other@example.com",
      role: "viewer",
      trustedOwnerKeyId: "key-2",
    };
    expect(profileDisplayLabel(state([shared]), shared)).toBe("EasyBC — other@example.com");
  });

  it("uses a device-local override without replacing the shared label", () => {
    const shared: ProfileRecord = {
      datasetId: "primary",
      ownerEmail: "other@example.com",
      folderName: "EasyBC — other@example.com",
      displayName: "Household plan",
      localDisplayName: "Alex",
      role: "viewer",
      trustedOwnerKeyId: "key-2",
    };
    expect(profileDisplayLabel(state([shared]), shared)).toBe("Alex");
  });

  it("adds owner email when visible labels conflict", () => {
    const a = { ...owned("a", "Plan"), ownerEmail: "a@example.com" };
    const b = { ...owned("b", "Plan"), ownerEmail: "b@example.com" };
    const value = state([a, b]);
    expect(disambiguatedProfileLabel(value, a)).toBe("Plan — a@example.com");
    expect(disambiguatedProfileLabel(value, b)).toBe("Plan — b@example.com");
  });

  it("labels local-only profiles independently of cloud ownership", () => {
    const local: ProfileRecord = {
      datasetId: "profile",
      ownerEmail: "local-123",
      folderName: "",
      displayName: "Offline journal",
      role: "owner",
      trustedOwnerKeyId: "",
      syncMode: "local",
    };
    expect(profileDisplayLabel(state([local]), local)).toBe("Offline journal");
  });

  it("keeps two owners' primary datasets distinct", () => {
    const local = owned("primary");
    const shared = {
      ...owned("primary"),
      ownerEmail: "other@example.com",
      folderName: "EasyBC — other@example.com",
      role: "writer" as const,
      trustedOwnerKeyId: "other-key",
    };
    const profiles = [local, shared];
    const value = state(profiles);
    expect(findProfile(value, "mom@example.com/primary")).toEqual(local);
    expect(findProfile(value, "other@example.com/primary")).toEqual(shared);
  });
});
